### PR TITLE
feat: Add `DeepSizeOf` decorator to `PeerManagerMessageRequest`

### PR DESCRIPTION
**Pull request by pmnoxx**
Hey, we are adding a big feature to around database. We will add a functionality to help investigate sizes of all data structures we use. This will require use to add `DeepSizeOf` to code that uses `BorshSerialize`. 

However, this has a lot of benefits:
- rate limiting - this will allow us to control memory usage - make sure our node doesn't ran out of memory
- debugging - it will be easy to check how big given data structure is, and optimize it by size
- network stats - we will be able to generate network stats. How big messages are, what's taking up memory / bandwidth. This will help us analyze, for example why nodes are out of sync sometimes.

This PR only adds `DeepSizeOf` decorator; this shouldn't effect current code execution in anyway.

This PR is  required to implement rate limiting, we want to control maximum memory usage in our node, per each connected peer. The goal is to stop reading from network, when we reach memory limits. https://github.com/near/nearcore/issues/5155
In the future, we will add more use cases this feature as well.

Split: 
- [x] part 1 - make `DeepSizeOf` decoration optional #5373

It will become optional in:
- near-network
- jsonrpc-adversarial-primitives
- near-network-primitives

### CODE REVIEW

**Comment by pmnoxx** [comment]
@bowenwang1996 @mm-near @matklad Can you take a look?

**Comment by pmnoxx** [comment]
@bowenwang1996 Unfortunately, I can't make this PR smaller.

**Comment by pmnoxx** [comment]
Check fails, because `deepsize_derive`  depends on `proc-macro2 0.4.30 `
While we use `proc-macro2 1.0.30`.

We may have to copy  `deepsize_derive`  or submit `PR` to them, but who knows how long will that take.

```
[2021-11-10T07:19:18Z] error[B004]: found 2 duplicate entries for crate 'proc-macro2'
--
  | [2021-11-10T07:19:18Z]     ┌─ /var/lib/buildkite-agent/builds/buildkite-i-0a8ad557d7118ea25-1/nearprotocol/nearcore/Cargo.lock:314:1
  | [2021-11-10T07:19:18Z]     │
  | [2021-11-10T07:19:18Z] 314 │ ╭ proc-macro2 0.4.30 registry+https://github.com/rust-lang/crates.io-index
  | [2021-11-10T07:19:18Z] 315 │ │ proc-macro2 1.0.30 registry+https://github.com/rust-lang/crates.io-index
  | [2021-11-10T07:19:18Z]     │ ```╰────────────────────────────────────────────────────────────────────────^ lock en

**Comment by matklad** [comment]
Note: we should be more careful when selecting dependencies. https://crates.io/crates/deepsize was updated last time a year ago and it is used by just two other crates. At the same time, it is of the "conquer the world" style of crates which requires annotating every type. So, this is a pretty risky dependency to add. 

Not saying that we shouldn't be adding it, but the PR with the addition should probably contain an explicit callout like "hey, we are adding this big capability to our codebase, it's somewhat experimental crate, but we feel its worth it". 

We should probably also review some alternatives as well. In particular, https://github.com/wasmerio/loupe is a more fresh alternative. There's also MallocSizeOf in the guts of servo. 

More generally, "getting sizes of things" in Rust is't a satisfactory solved problem at the moment. I wanted to have a solution for this for ages (fun fact, the deepsize crate was created upon my request to solve this problem https://github.com/dtolnay/request-for-implementation/issues/22), but, eg, in rust-analyzer we still don't use that, as it isn't worth it, and instead do [horrible tricks](https://rust-analyzer.github.io/blog/2020/12/04/measuring-memory-usage-in-rust.html). I do think that, if we want exactly this, than taking on deepsize dependency is probably the way to go, it's just that this'll add non-zero long-term maintenance.

I am not sure we want exactly this though. My understanding is that the end consumer here would be rate limiting, and for that, we want to estimate a size of message. My gut feeling is that just writing a `fn messge_size(msg: &Msg) -> usize` by hand would probably be a more robust approach: 

* we can write it in one place, without changes to the codebase elsewhere
* with "one place", it will be much easier to tweak/remove/replace
* we care about approximate answer, and, ideally, we'd take not just the size of the message, but the amount of memory it takes to process. With hand-written function, it would be much easier to do tweaks like "multiply the size of this particular kind of message by two, as they are costly to process".

**Comment by matklad** [comment]
Note: if we do take on this depenrency, we should send a PR upstream to upgrade its `syn` version. Compiling two versions of `syn` is not fun.

**Comment by pmnoxx** [comment]
I checked dependencies of `deepsize_derive`. In https://github.com/Aeledfyr/deepsize/blob/master/deepsize_derive/Cargo.toml
they specified to use stable version of libraries:
```
[package]
name = "deepsize_derive"
version = "0.1.1"
...

[dependencies]
proc-macro2 = "^1.0"
syn = "^1.0"
quote = "^1.0"
```

However, in our `Cargo.lock`, I see that the old ones are used

```

[[package]]
name = "deepsize_derive"
version = "0.1.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "eefac8b754bebccb018f15a8ff0015a3f58fb298825f5bd0dcf3e7e2c017e992"
dependencies = [
 "proc-macro2 0.4.30",
 "quote 0.6.13",
 "syn 0.15.44",
]
```

@matklad Any ideas?


I checked out `https://github.com/wasmerio/loupe`. It was updated 8 months ago version 13 months ago for `DeepSizeOf`. So it's not much different. I think that `DeepSizeOf` is better maintained, it supports more types. Anyway, I couldn't find `MallocSizeOf` that you mentioned.

**Comment by pmnoxx** [comment]
> Note: we should be more careful when selecting dependencies. https://crates.io/crates/deepsize was updated last time a year ago and it is used by just two other crates. At the same time, it is of the "conquer the world" style of crates which requires annotating every type. So, this is a pretty risky dependency to add.

Yes, I agree. Though, I think that the benefits outweigh the drawbacks. In the near future, we should implement adding network stats. I would like to see how big are all messages that we send over the network. That can help us explain which data structures are big, and consume a lot of bandwidth, use a lot of memory etc.

> Not saying that we shouldn't be adding it, but the PR with the addition should probably contain an explicit callout like "hey, we are adding this big capability to our codebase, it's somewhat experimental crate, but we feel its worth it".

Good idea, I updated the PR.

> We should probably also review some alternatives as well. In particular, https://github.com/wasmerio/loupe is a more fresh alternative. There's also MallocSizeOf in the guts of servo.

I looked it, but currently it offer less features than `DeepSizeOf`. However, I'm open to suggestions I can replace `DeepSizeOf` with any other library if needed.
 
> More generally, "getting sizes of things" in Rust is't a satisfactory solved problem at the moment. I wanted to have a solution for this for ages (fun fact, the deepsize crate was created upon my request to solve this problem [dtolnay/request-for-implementation#22](https://github.com/dtolnay/request-for-implementation/issues/22)), but, eg, in rust-analyzer we still don't use that, as it isn't worth it, and instead do [horrible tricks](https://rust-analyzer.github.io/blog/2020/12/04/measuring-memory-usage-in-rust.html). I do think that, if we want exactly this, than taking on deepsize dependency is probably the way to go, it's just that this'll add non-zero long-term maintenance.

Yes, we should investigate if there is a better way to do it. However, I think this is a good enough, if we find a better way we will replace DeepSizeOf with something equivalent.
 
> I am not sure we want exactly this though. My understanding is that the end consumer here would be rate limiting, and for that, we want to estimate a size of message. My gut feeling is that just writing a `fn messge_size(msg: &Msg) -> usize` by hand would probably be a more robust approach:

This is not the only use case. Previously, when I was investigating memory leaks, I used `DeepSizeOf` to print sizes of data structures to see how big they are. Right now, it's hard to estimate how big our data structures are by just looking at the code.
Sometimes, you want to check whenever a particular data structure is big enough, that it should be optimized or not.
 
> * we can write it in one place, without changes to the codebase elsewhere
> * with "one place", it will be much easier to tweak/remove/replace
> * we care about approximate answer, and, ideally, we'd take not just the size of the message, but the amount of memory it takes to process. With hand-written function, it would be much easier to do tweaks like "multiply the size of this particular kind of message by two, as they are costly to process".

Yes, that's possible. However, writing manually code for over 500 data structures would be both time consuming, and hard to maintain. Changing any data structure, would require changes in two places. Hand writing code with any estimator my by really error-prone, requiring extra time. I strongly feel having to write manual method each time would add a lot of maintenance.

**Comment by matklad** [comment]
>I couldn't find MallocSizeOf that you mentioned.

Oups, sorry, should have added a link: https://github.com/servo/servo/tree/master/components/malloc_size_of. It implements things slightly differently -- rather than using `size_of`, it asks the allocator about the size of the underlying allocation, so:

* it is more precise, as it accounts for extra padding when the object doesn't fit exactly into allocator's size classes
* it is more fragile, as it requires runtime-support from a particular allocator. 

>Previously, when I was investigating memory leaks, I used DeepSizeOf to print sizes of data structures to see how big they are. 

Yeah, that' very convincing argument for having this thing. 

> Any ideas?

The code on github wasn't published yet, I've asked the author to publish: https://github.com/Aeledfyr/deepsize/commit/ffc79ef6277d3e14a6b5ced49a6f45e825c6c36f#r60105545

File: chain/jsonrpc-adversarial-primitives/src/lib.rs
Comment (pmnoxx) [reply chain/jsonrpc-adversarial-primitives/src/lib.rs:12]: `json-adversarial-primitives` is only enabled when code is compiled with `test_features`.

File: core/crypto/src/signature.rs
Comment (pmnoxx) [reply core/crypto/src/signature.rs]: Fixed. You are right, this array will not cause an allocation, and the size should be 0.

**Comment by pmnoxx** [comment]
> * it is more fragile, as it requires runtime-support from a particular allocator.

Yes, it may be worth considering it. However, doing c calls may have some performance implications, because that code can't be fully optimized by the compiler. We would have to compare the performance.

It's probably fine for us to use less precise measurement. 

I'm also a little bit worried that this may not work on all platforms or with some memory allocators.

File: chain/network-primitives/src/types.rs
Comment (pmnoxx) [reply chain/network-primitives/src/types.rs:819]: Yeah, I was thinking about doing that.

File: chain/network/src/types.rs
Comment (pmnoxx) [reply chain/network/src/types.rs:453]: Yeah, I was thinking about doing that.

Comment (pmnoxx) [reply chain/network/src/types.rs:453]: https://github.com/Aeledfyr/deepsize/pull/18

File: chain/network-primitives/src/types.rs
Comment (pmnoxx) [reply chain/network-primitives/src/types.rs:819]: https://github.com/Aeledfyr/deepsize/pull/18

**Comment by pmnoxx** [comment]
@matklad We may have to create a work for `DeepSizeOf`. We don't know how much time it will take for the author to publish new create with updated dependency, and that's blocking us.

**Comment by matklad** [comment]
deepsize-derive with syn 1.0 was published just yesterday, that's great: https://crates.io/crates/deepsize-derive/0.1.2

**Comment by pmnoxx** [comment]
@matklad Yes, but  we have to wait for `deepsize`, right?

**Comment by matklad** [comment]
Yeah, I mean that in general the maintainer is responsive, so we can hope for deep size publishing.

**Comment by frol** [comment]
This change is taxing all the crates. I hate to be this guy, but at the very least we need to make this dependency optional.

**Comment by pmnoxx** [comment]
> This change is taxing all the crates. I hate to be this guy, but at the very least we need to make this dependency optional.

@frol No problem. I can make it so `DeepSize` trait is noop as long as you don't compile `near-network.`

That's actually a good idea. I can implement the change #5198
without usage of DeepSizeOf. Even without `DeepSizeOf` we will be able to get most of needed functionality. That will allow us to unblock #5198

Later, we will be able to decide whenever we want it always enabled. We can do it just for the purpose of `near-network`.

**Comment by pmnoxx** [comment]
> This change is taxing all the crates. I hate to be this guy, but at the very least we need to make this dependency optional.

Btw, what are your main concerns?
- compilation time
- extra memory - deepsize doesn't increase size of structs
- maintenance - having to add `DeepSize` to all structs
- crate size - this crate has very little code, it doesn't be an issue.
- adding extra  dependencies for other binaries - like indexer - I agree, `indexer` and other binaries don't need it.

**Comment by pmnoxx** [comment]
@frol Here is how we can do it without adding extra import.

We could do it for other derives, and/or traits as well if we want to.

**Comment by pmnoxx** [comment]
@frol @matklad @mm-near Can you take a look?

We merged https://github.com/near/nearcore/pull/5373 where we make `DeepSizeOf` optional for most crates.

This PR makes `DeepSizeOf` required for 
- `jsonrpc-adversarial-primitives`
- `near-network`
- `near-network-primitives`

This PR is ready. The only last question is to decide whenever we want to make `jsonrpc-adversarial-primitives` and ` `near-network-primitives` use `deepsize` as an optional dependency.

File: chain/jsonrpc-adversarial-primitives/Cargo.toml
Comment (matklad) [inline chain/jsonrpc-adversarial-primitives/Cargo.toml:24]: Sadly, I don't know a way to make `deepsize` name here "just work".

**Review by matklad — APPROVED** [review]

File: chain/jsonrpc-adversarial-primitives/Cargo.toml
Comment (pmnoxx) [reply chain/jsonrpc-adversarial-primitives/Cargo.toml:24]: @matklad Yes, that's unfortunate.

File: chain/network/src/types.rs
Comment (mm-near) [inline chain/network/src/types.rs:472]: What is *this* that you're referring to in 'we can't implement this'?

**Review by mm-near — APPROVED** [review]

File: chain/network-primitives/Cargo.toml
Comment (frol) [inline chain/network-primitives/Cargo.toml:21]: I feel we should not enable the feature unconditionally:

```suggestion
near-primitives = { path = "../../core/primitives" }
```

I see the same pattern in other Cargo.toml files, so we should address it there as well

File: chain/network/src/routing/ibf_peer_set.rs
Comment (frol) [inline chain/network/src/routing/ibf_peer_set.rs:4]: This should be behind a feature-flag.

P.S. I feel it would be easier to just use a fully-qualified path inside the `derive()` to avoid a need to have a conditional use.

**Review by frol — APPROVED** [review]
LGTM! Make your own judgement whether it makes sense to address my review comments

File: chain/network/src/types.rs
Comment (pmnoxx) [reply chain/network/src/types.rs:472]: I replaced this comment with a more descriptive one
```
/// Addr<PeerActor> doesn't implement `DeepSizeOf` waiting for `deepsize` > 0.2.0.
```

File: chain/network-primitives/Cargo.toml
Comment (pmnoxx) [reply chain/network-primitives/Cargo.toml:21]: Thanks. Fixed.

File: chain/network/src/routing/ibf_peer_set.rs
Comment (pmnoxx) [reply chain/network/src/routing/ibf_peer_set.rs:4]: > This should be behind a feature-flag.

@frol `near-network` requires `deepsize::DeepSizeOf` to work. That's why it's not under feature flag. However, we may make `near-network` optionally require `deepsize`, but I think this should be addressed by another PR.

> P.S. I feel it would be easier to just use a fully-qualified path inside the `derive()` to avoid a need to have a conditional use.

I agree, this was suggested by @matklad

**Comment by pmnoxx** [comment]
I'll prepare another PR, which deals with resolves issue of conditional imports.

File: chain/network/src/routing/ibf_peer_set.rs
Comment (pmnoxx) [reply chain/network/src/routing/ibf_peer_set.rs]: @frol Oh, I realized that I made all packages use `deepsize_feature` optionally. 
Thanks I fixed it.

I also did a PR, where I do it for whole `near-network`: https://github.com/near/nearcore/pull/5436

**Pull request merged**
